src/util/invariant.cpp

```diff
@@ -0,0 +1,154 @@
+/*******************************************************************\
+
+Module:
+
+Author: Martin Brain, martin.brain@diffblue.com
+
+\*******************************************************************/
+
+
+#include "invariant.h"
+
+#include <stdexcept>
+#include <string>
+#include <sstream>
+
+#ifdef CPROVER_INVARIANT_PRINT_STACK_TRACE
+#include <iostream>
+#include <cstdlib>
+#endif
+
+// Backtraces compiler and C library specific
+// So we should include something explicitly from the C library
+// to check if the C library is glibc.
+#include <assert.h>
+#ifdef __GLIBC__
+
+// GCC needs LINKFLAGS="-rdynamic" to give function names in the backtrace
+#include <execinfo.h>
+#include <cxxabi.h>
+
+
+/*******************************************************************\
+
+Function: output_demangled_name
+
+  Inputs: The output stream and the description of the stack_entry
+
+ Outputs: True <=> the entry has been successfully demangled and printed.
+
+ Purpose: Attempts to demangle the function name assuming the glibc
+          format of stack entry:
+
+          path '(' mangled_name '+' offset ') [' address ']\0'
+
+\*******************************************************************/
+
+static bool output_demangled_name(
+  std::ostream &out,
+  const char * const stack_entry)
+{
+  bool return_value=false;
+
+  std::string working(stack_entry);
+
+  std::string::size_type start=working.rfind('(');  // Path may contain '(' !
+  std::string::size_type end=working.find('+', start);
+
+  if(start!=std::string::npos &&
+     end!=std::string::npos &&
+     start+1<=end-1)
+  {
+    std::string::size_type length=end - (start+1);
+    std::string mangled(working.substr(start+1, length));
+
+    int demangle_success=1;
+    char *demangled=
+      abi::__cxa_demangle(mangled.c_str(), NULL, 0, &demangle_success);
+
+    if(demangle_success==0)
+    {
+      out << working.substr(0, start+1)
```

Comment:
If the ostream has exceptions enabled, `<<` may throw an exception, in which case we would leak `demangled`.